src/main/kotlin/com/mdrobnak/lalrpop/codeInsight/LpExpressionTypeInlayHintProvider.kt

```diff
@@ -0,0 +1,126 @@
+package com.mdrobnak.lalrpop.codeInsight
+
+import com.intellij.codeInsight.hints.*
+import com.intellij.codeInsight.hints.ImmediateConfigurable.Case
+import com.intellij.openapi.components.service
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.DumbService
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.parentOfType
+import com.mdrobnak.lalrpop.injectors.findModuleDefinition
+import com.mdrobnak.lalrpop.psi.*
+import com.mdrobnak.lalrpop.psi.ext.importCode
+import com.mdrobnak.lalrpop.psi.ext.selected
+import org.rust.ide.hints.type.RsTypeHintsPresentationFactory
+import org.rust.lang.core.psi.ext.endOffset
+import javax.swing.JComponent
+import javax.swing.JPanel
+
+@Suppress("UnstableApiUsage")
+class LpExpressionTypeInlayHintProvider : InlayHintsProvider<LpExpressionTypeInlayHintProvider.Settings> {
+    class Settings(
+        var showForNonterminals: Boolean = true,
+        var showForSymbols: Boolean = true,
```

Comment:
This setting might want to be disabled by default. It really expands the line and sometimes isn't too useful. It would be really great to instead see the type on hover (default is on hover, can explicitly enable this inlay hint as well), but that might be for a later PR.